designs/2024-report-unused-inline-configs/README.md

```diff
@@ -0,0 +1,253 @@
+For a rough code reference, see [`poc: reporting unused inline configs`](https://github.com/JoshuaKGoldberg/eslint/commit/e14e404ed93e6238bdee817923a449f5215eecd8).
+
+### Computing Option Differences
+
+Each inline config comment will be compared against the existing configuration value(s) it attempts to override:
+
+- If the config comment only specifies a severity, then only the severity will be checked for redundancy
+  - The new logic will normalize options: `"off"` will be considered equivalent to `0`
+- If the config comment also specifies rule options, they will be compared for deep equality to the existing rule options
+
+This RFC should wait to begin work until after [feat: add meta.defaultOptions](https://github.com/eslint/eslint/pull/17656) is merged into ESLint.
+That way, a rule's `meta.defaultOptions` can be factored into computing whether an inline config's rule options differ from the previously configured options.
+
+### Default Values
+
+This RFC proposes a two-step approach to introducing unused inline config reporting:
+
+1. In the current major version upon introduction, don't enable unused inline config reporting by default
+2. In the next major version, enable unused inline config reporting by default
+
+Note that the default value in the next major version should be the same as reporting unused disable directives.
+See [Change Request: error by default for unused disable directives](https://github.com/eslint/eslint/issues/18665) for an issue on changing that from `"warn"` to `"error"`.
+
+## Documentation
+
+The new settings will be documented similarly to reporting unused disable directives:
+
+- [Configuration Files](https://eslint.org/docs/latest/use/configure/configuration-files):
+  - List item for `reportUnusedInlineConfig` under _[Configuration Objects](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects)_ > `linterOptions`
+  - Sub-heading alongside _[Reporting Unused Disable Directives](https://eslint.org/docs/latest/use/configure/configuration-files#reporting-unused-disable-directives)_
+- [Configure Rules](https://eslint.org/docs/latest/use/configure/rules):
+  - Sub-heading alongside _[Report unused eslint-disable comments](https://eslint.org/docs/latest/use/configure/rules#report-unused-eslint-disable-comments)_
+- [Command Line Interface Reference](https://eslint.org/docs/latest/use/command-line-interface):
+  - List item under the _[Options](https://eslint.org/docs/latest/use/command-line-interface#options)_ code block, under `Inline Configuration comments:`
+  - Corresponding sub-headings under _[Inline Configuration Comments](https://eslint.org/docs/latest/use/command-line-interface#inline-configuration-comments)_
+
+## Drawbacks
+
+Any added options come with an added tax on project maintenance and user comprehension.
```

Comment:
What about performance? Do you have an idea of this impact?

For context, reporting unused disable directives is fairly lightweight, especially if there are no linting errors, so we were okay to enable that by default.

Reply:
Yeah, it should be the same here. Nothing goes beyond O(1) and there are rarely more than a handful of config comments in a file. I'll add an analysis section.